README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
emyweather
==========

[![Travis-CI Build Status](https://travis-ci.org/kvantas/emyweather.svg?branch=master)](https://travis-ci.org/kvantas/emyweather) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/kvantas/emyweather?branch=master&svg=true)](https://ci.appveyor.com/project/kvantas/emyweather) [![minimal R version](https://img.shields.io/badge/R%3E%3D-3.4.0-6666ff.svg)](https://cran.r-project.org/) [![packageversion](https://img.shields.io/badge/Package%20version-0.2.0-orange.svg?style=flat-square)](https://github.com/kvantas/emyweather) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1159500.svg)](https://doi.org/10.5281/zenodo.1159500)

`emyweather` is an R package that contains climatic data from the [Greek National Meteorological Service](http://emy.gr), as well as a number of useful other related data sets. The climatic data come from 21 weather stations, with a time step of 3 hours and the time period is from 1975 to 2004. For more details checkout the package's [website](https://kvantas.github.io/emyweather/).

<img src="https://github.com/kvantas/emyweather/raw/master/man/figures/stations_map.png" align = "center"/>

Installation
------------

You can install emyweather from github with:

``` r
# install.packages("devtools")
devtools::install_github("kvantas/emyweather")
```

Using emyweather
----------------

This package is comprised of three data sets:

-   `emy_data`: climatic measurements
-   `emy_stations`: weather stations' coordinates and elevation
-   `borders`: borders of Greece

### `emy_data`

The variables in this data set are:

| Variable             | Description                                      |
|----------------------|--------------------------------------------------|
| date                 | The observation's date                           |
| station\_id          | Greek National Meteorological Service station ID |
| pressure             | Pressure, in hPa                                 |
| dry\_temp            | Dry temperature, in degree Celsius               |
| wet\_temp            | Wet temperature, in degree Celsius               |
| max\_temp            | Maximum temperature, in degree Celsius           |
| min\_temp            | Minimum temperature, in degree Celsius           |
| rel\_humidity        | Relative humidity, in percent                    |
| wind\_direction      | Wind direction, in degrees                       |
| wind\_force\_knot    | Wind force, in knots                             |
| visibility           | Visibility, in km                                |
| total\_cloud\_amount | Total cloud amount, in values 0 to 9             |
| precipitation        | Precipitation height, in mm                      |

Please note that these time series are not checked for consistency or errors.

### `emy_stations`

The stations' data are taken from the Greek National Data Bank for Hydrological and Meteorological Information, [Hydroscope](http://www.hydroscope.gr/) using the package [`hydroscoper`](https://github.com/ropensci/hydroscoper/blob/master/vignettes/intro_hydroscoper.Rmd). The variables are:

| Variable        | Description                                        |
|-----------------|----------------------------------------------------|
| name            | The station's name                                 |
| station\_id     | Greek National Meteorological Service station ID   |
| hydroscope\_id  | The station's ID from kyy.hydroscope.gr database   |
| water\_division | The station's Water Division                       |
| longitude       | The station's longitude in decimal degrees, ETRS89 |
| latitude        | The station's latitude in decimal degrees, ETRS89  |
| altitude        | The station's altitude, meters above sea level     |

### `borders`

The borders of Greece are taken from Geoadata.gov.gr. The variables are created using the function `tidy` from the [`broom`](https://cran.r-project.org/web/packages/broom/index.html) package for use with `geom_polygon` from [`ggplot2`](https://cran.r-project.org/web/packages/ggplot2/).

Meta
----

-   Please [report any issues or bugs](https://github.com/kvantas/emyweather/issues).
-   Licence:
    -   All code is licensed MIT.
    -   All data are from the open data sources in <http://www.emy.gr/>.
-   To cite `emyweather` in publications, please use:

          Vantas, (2018). emyweather: R package with open climatic data from the Greek National
          Meteorological Service, URL: https://github.com/kvantas/emyweather,
          DOI:http://doi.org/10.5281/zenodo.1159500

    A BibTeX entry for LaTeX users is

        @Manual{ vantas2018emy,
        author = {Konstantinos Vantas},
        title = {{emyweather}: R package with open climatic data from the Greek National Meteorological Service},
        doi = {10.5281/zenodo.1159500},
        year = {2018},
        note = {R package version 0.2.0},
        url = {https://github.com/kvantas/emyweather},
          }

-   Please note that this project is released with a [Contributor Code of Conduct](/CONDUCT.md). By participating in this project you agree to abide by its terms.

References
----------

-   [GEODATA.gov.gr](http://geodata.gov.gr/)
-   [Hydroscope](http://www.hydroscope.gr/)
-   [National Meteorological Service](http://www.emy.gr/)


NEWS.md
# emyweather 0.2.0 (Release date: 2018-03-17)
--------------------------------------------------------------------------------

* Convert data sets to tibbles.

# emyweather 0.1.0 (Release date: 2018-01-25)
--------------------------------------------------------------------------------

* Initial version.





R/emy_data.R
#' Climatic data from the Greek National Meteorological Service
#'
#' The climatic data are taken from the Greek National Meteorological Service
#' and come from 21 weather stations, with a time step of 3 hours. The time
#' period is from 1975 to 2004. These time series were not checked for
#' consistency or errors.
#'
#' @format A data frame with 1679308 rows and 13 variables:
#' \describe{
#'   \item{date}{the observation's date}
#'   \item{station_id}{Greek National Meteorological Service station ID}
#'   \item{pressure}{Pressure, in hPa}
#'   \item{dry_temp}{Dry temperature, in degree Celsius}
#'   \item{wet_temp}{Wet temperature, in degree Celsius}
#'   \item{max_temp}{Maximum temperature, in degree Celsius}
#'   \item{min_temp}{Minimum temperature, in degree Celsius}
#'   \item{rel_humidity}{Relative humidity, in percent}
#'   \item{wind_direction}{Wind direction, in degrees}
#'   \item{wind_force_knot}{Wind force, in knots}
#'   \item{visibility}{Visibility, in km}
#'   \item{total_cloud_amount}{Total cloud amount, in values 0 to 9}
#'   \item{precipitation}{Precipitation height, in mm}
#' }
#' @source \url{http://www.emy.gr/}
"emy_data"


data-raw/emy_create.R

#' Create National Meteorological Service Dataset
#'
#' \code{emy_create} uses as input a list of directories. Each directory is a
#' station and the files in it are the open data, csv files, from www.emy.gr.
#'
#' @param dir_names is a directories list.
#'
#' @return a tidy data frame with the open meteorological data from National
#' Meteorological Service
#' @export emy_create
#'
#' @examples
#' \dontrun{
#' # read a list of directories in ./data/emy
#' dir_names <- list.files('./data-raw/emy', full.names = TRUE)
#'
#' # create a data frame using all the files in the directories list
#' emy_data <- emy_create(dir_names)
#' }
emy_create <- function(dir_names) {

  plyr::ldply(dir_names, function(directory) {
    message(paste("Parsing ", directory))
    files <- list.files(directory, full.names = TRUE)

    data <- plyr::ldply(files, function(in_file){

      ## read file
      tmp <- utils::read.table(file = in_file,
                               header = TRUE,
                               stringsAsFactors = FALSE,
                               sep = ";",
                               dec = ",")

      ## create Dates
      tmp$Date <- paste0(tmp$OBSERVATIONYEAR, "-",
                         tmp$OBSERVATIONMONTH, "-",
                         tmp$OBSERVATIONDAY)
      ## create Time
      tmp$Time <- formatC(tmp$OBSERVATIONHOUR, width = 2, format = "d",
                          flag = "0")
      tmp$Time  <- paste0(tmp$Time, ":00")

      ## create POSIXlt
      tmp$DT <- paste(tmp$Date, tmp$Time)
      tmp$DTposix <- as.POSIXct(tmp$DT, format =  "%Y-%m-%d %H:%M",
                               tz = "Europe/Athens")

      ## convert pressure to numeric
      tmp$BARPRESSURE <- stringr::str_replace(tmp$BARPRESSURE, "\\.", "")
      tmp$BARPRESSURE <- stringr::str_replace(tmp$BARPRESSURE, "\\,", "\\.")
      tmp$BARPRESSURE <- as.numeric(tmp$BARPRESSURE)

      ## translate names
      tmp_sub <- tmp[c("DTposix", "STATIONID", "BARPRESSURE", "DRYTEMP",
                       "WETTEMP", "MAXTEMP", "MINTEMP", "RELHUMIDITY",
                       "WINDDIRECTION", "WINDFORCEKNOT", "VISIBILITY",
                       "TOTALCLOUDAMOUNT", "PRECIPHEIGHT")]
      names(tmp_sub) <- tolower(names(tmp_sub))
      names(tmp_sub) <- c("date", "station_id", "pressure", "dry_temp",
                          "wet_temp", "max_temp", "min_temp", "rel_humidity",
                          "wind_direction", "wind_force_knot", "visibility",
                          "total_cloud_amount", "precipitation")
      return(tmp_sub)
    })
  })
}


R/emyweather-package.R
#' emyweather: Open data from the Greek National Meteorological Service
#'
#' This package contains climatic data from the Greek National
#' Meteorological Service, as well as a number of useful other related data
#' sets.
#'
#' @section Climatic Data:
#'
#' The climatic data are taken from the Greek National Meteorological Service
#' and come from 21 weather stations, with a time step of 3 hours. The time
#' period is from 1975 to 2004. The variables are: air pressure, temperature
#' (dry, wet, max and min), relative humidity, wind (direction and speed),
#' visibility, total cloud amount and precipitation. These time series were not
#' checked for consistency or errors.
#'
#' @section Stations:
#'
#' The stations' data are taken from the Greek National Data Bank for
#' Hydrological and Meteorological Information, Hydroscope. The variables are:
#' stations' id, coordinates, altitude and water division.
#'
#' @section Borders of Greece:
#'
#' The variables are created using the function \code{tidy} from the
#' \code{broom} package for use with the \code{ggplot2} package.
#'
#' @section Data Sources:
#'
#' The data are retrieved from:
#' \itemize{
#'  \item \href{http://www.emy.gr}{Greek National Meteorological Service}
#'  \item \href{http://www.hydroscope.gr/}{Hydroscope}
#'  \item \href{http://geodata.gov.gr/}{GEODATA.gov.gr}
#' }
#' @name emyweather
#' @aliases emyweather-package
#' @docType package
"_PACKAGE"


data-raw/create_borders.R
# load packages
library(ggplot2)
library(broom)
library(rgdal)
library(ggmap)

#load greece polygon
borders <- readOGR(dsn = "./data-raw/greece_shp", layer = "greece")
proj4string(borders) = CRS("+init=epsg:2100")

#convert to ETRS89
etrs89 <- CRS("+init=epsg:4258")
borders <- spTransform(borders, etrs89)

# convert shp file to dataframe
borders <- broom::tidy(borders)

# remove id variable (no values in there)
borders$id <- NULL

map <-
  get_map(
    location = c(lon = mean(borders$lon), mean(borders$lat)),
    color = 'bw',
    zoom = 6,
    maptype = "watercolor",
    language = 'greek'
  )
p <- ggmap(map)


# plot data
p+
  geom_polygon( data = borders,
    aes(x = long, y = lat, group = group),
    color = "black",
    size = 0.5,
    alpha = 0.1,
    fill = NA
  ) +
  geom_point(
    data = emy_stations,
    aes(x = longitude, y = latitude),
    color = "black",
    fill = "red",
    shape = 21,
    size = 3
  ) + theme_bw()+
  xlim(min(borders$lon), max(borders$lon)) +
  ylim(min(borders$lat), max(borders$lat))+
  labs(x = "Longitude", y = "Latitude")

# save data
borders <- tibble::as.tibble(borders)
devtools::use_data(borders, overwrite = TRUE)


data-raw/create_timeser_df.R

#' Greek National Meteorological Service Dataset
#'
#' \code{emy_create} uses as input a list of directories. Each directory is a
#' station and the files in it are the open data, csv files, from www.emy.gr.
#'
#' @param dir_names is a directories list.
#'
#' @return a tidy data frame with the open meteorological data from National
#' Meteorological Service
#' @export emy_create
#'
#' @examples
#' \dontrun{
#' # read a list of directories in ./data/emy
#' dir_names <- list.files('./data-raw/emy', full.names = TRUE)
#'
#' # create a data frame using all the files in the directories list
#' emy_data <- emy_create(dir_names)
#' }
emy_create <- function(dir_names) {

  plyr::ldply(dir_names, function(directory) {
    message(paste("Parsing ", directory))
    files <- list.files(directory, full.names = TRUE)

    data <- plyr::ldply(files, function(in_file){

      ## read file
      tmp <- utils::read.table(file = in_file,
                               header = TRUE,
                               stringsAsFactors = FALSE,
                               sep = ";",
                               dec = ",")

      ## create Dates
      tmp$Date <- paste0(tmp$OBSERVATIONYEAR, "-",
                         tmp$OBSERVATIONMONTH, "-",
                         tmp$OBSERVATIONDAY)
      ## create Time
      tmp$Time <- formatC(tmp$OBSERVATIONHOUR, width = 2, format = "d",
                          flag = "0")
      tmp$Time  <- paste0(tmp$Time, ":00")

      ## create POSIXlt
      tmp$DT <- paste(tmp$Date, tmp$Time)
      tmp$DTposix <- as.POSIXct(tmp$DT, format =  "%Y-%m-%d %H:%M",
                               tz = "Europe/Athens")

      ## convert pressure to numeric
      tmp$BARPRESSURE <- stringr::str_replace(tmp$BARPRESSURE, "\\.", "")
      tmp$BARPRESSURE <- stringr::str_replace(tmp$BARPRESSURE, "\\,", "\\.")
      tmp$BARPRESSURE <- as.numeric(tmp$BARPRESSURE)

      ## translate names
      tmp_sub <- tmp[c("DTposix", "STATIONID", "BARPRESSURE", "DRYTEMP",
                       "WETTEMP", "MAXTEMP", "MINTEMP", "RELHUMIDITY",
                       "WINDDIRECTION", "WINDFORCEKNOT", "VISIBILITY",
                       "TOTALCLOUDAMOUNT", "PRECIPHEIGHT")]
      names(tmp_sub) <- tolower(names(tmp_sub))
      names(tmp_sub) <- c("date", "station_id", "pressure", "dry_temp",
                          "wet_temp", "max_temp", "min_temp", "rel_humidity",
                          "wind_direction", "wind_force_knot", "visibility",
                          "total_cloud_amount", "precipitation")
      return(tmp_sub)
    })
  })
}


R/data.R
#' Climatic data from the Greek National Meteorological Service
#'
#' The climatic data are taken from the Greek National Meteorological Service
#' and come from 21 weather stations, with a time step of 3 hours. The time
#' period is from 1975 to 2004. These time series were not checked for
#' consistency or errors.
#'
#' @format A tibble with 1679308 rows and 13 variables:
#' \describe{
#'   \item{date}{The observation's date}
#'   \item{station_id}{Greek National Meteorological Service station ID}
#'   \item{pressure}{Pressure, in hPa}
#'   \item{dry_temp}{Dry temperature, in degree Celsius}
#'   \item{wet_temp}{Wet temperature, in degree Celsius}
#'   \item{max_temp}{Maximum temperature, in degree Celsius}
#'   \item{min_temp}{Minimum temperature, in degree Celsius}
#'   \item{rel_humidity}{Relative humidity, in percent}
#'   \item{wind_direction}{Wind direction, in degrees}
#'   \item{wind_force_knot}{Wind force, in knots}
#'   \item{visibility}{Visibility, in km}
#'   \item{total_cloud_amount}{Total cloud amount, in values 0 to 9}
#'   \item{precipitation}{Precipitation height, in mm}
#' }
#' @source \url{http://www.emy.gr/}
"emy_data"

#' Greek National Meteorological Service Stations
#'
#' The stations' data are taken from the Greek National Data Bank for
#' Hydrological and Meteorological Information, Hydroscope. The variables are:
#' stations' id, coordinates, altitude and water division.
#'
#' @format A tibble with 21 rows and 7 variables:
#' \describe{
#'   \item{name}{The station's name}
#'   \item{station_id}{Greek National Meteorological Service station ID}
#'   \item{hydroscope_id}{The station's ID from kyy.hydroscope.gr database}
#'   \item{water_division}{The station's Water Division}
#'   \item{longitude}{The station's longitude in decimal degrees, ETRS89}
#'   \item{latitude}{The station's latitude in decimal degrees, ETRS89}
#'   \item{altitude}{The station's altitude, meters above sea level}
#' }
#' @source \url{http://www.hydroscope.gr/}
"emy_stations"


#' Greek borders
#'
#' The borders of Greece are taken from geodata.gov.gr.  The variables are
#' created using the function tidy from the broom package. This data frame
#' was created for use with the geom_polygon from ggplot2 package.
#'
#' @format A tibble with 18474 rows and 8 variables:
#' \describe{
#'   \item{long}{Longitude in decimal degrees, ETRS89}
#'   \item{lat}{Latitude in decimal degrees, ETRS89}
#'   \item{order}{order, integer}
#'   \item{hole}{hole, boolean}
#'   \item{piece}{piece, integer}
#'   \item{group}{group, numeric}
#' }
#' @source \url{http://geodata.gov.gr/}
"borders"
